### DISCUSSION TITLE
Poetry is taking first wheel metadata only and thus fails to find package

### DISCUSSION THREAD

**Discussion by alexis-girard** [python-poetry/poetry#8531] [Q&A] [answered]
Hi everyone, 

I didn't know where I should be posting this question, we have a python library packaged and uploaded on an internal repository (Artifactory).
This library is built for a range of Python 3.6-3.10 and is based on numpy. 
As you may know, numpy compatibility is dependant of Python version.
For example, python 3.6 is compatible with this range of numpy 1.17.0 - 1.20.0 while python 3.10 is compatible with numpy >= 1.22

With poetry using the first wheel it finds to extract the metadata, it fails to find the package version compatible with Python 3.6/numpy 1.17

Here is an example :

```
[tool.poetry.dependencies]
python = "3.6.8"
numpy = "1.17.0"
my-package= "1.0.0"
```

Here is my poetry env info

```
Virtualenv
Python:         3.6.8
Implementation: CPython
Path:           ...
Executable:     ...
Valid:          True

System
Platform:   win32
OS:         nt
Python:     3.6.8
Path:       C:\Python\3.6.8
Executable: C:\Python\3.6.8\python.exe
```

The first package in the list is the package for Python 3.10 and poetry reports a version solving failed due to numpy requirement not satisfied.
 
Am I not using correctly ? How am I supposed to handle this, any suggestions are welcome.

Best,
Alexis.

**Comment by radoering** [comment, marked as answer]
All your wheels should have the same requirements using environment markers like
```
numpy >=1.17, <1.18 ; python_version == "3.6"
numpy >=1.18, <1.20 ; python_version == "3.7"
...

```

**Comment by jyligehc** [comment]
I do have similar problem with my packages.

I have a two package that requires specific CPython version (Actually 4, two per specific OS):
These are created with Poetry and uploaded Artifactory PyPi:
`mypkg-7.2.1-cp310-none-win_amd64.whl` (inside package requires-python >=3.10,< 3.11)
`mypkg-7.2.1-cp311-none-win_amd64.whl` (inside package requires-python >=3.11,< 3.12)

In the main pyproject:
```
[tool.poetry.dependencies]
python = ">=3.10,<3.12"
mypkg = { version = "7.2.1", python = ">=3.10,<3.12"}
```
executing poetry lock:
```
The current project's supported Python range (>=3.10,<3.12) is not compatible with some of the required packages Python requirement:
  - mypkg requires Python >=3.10,<3.11, so it will not be satisfied for Python >=3.11,<3.12

 Because taf depends on mypkg (7.2.1) which requires Python >=3.10,<3.11, version solving failed.
  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    For squish, a possible solution would be to set the `python` property to ">=3.10,<3.11"
```

Why isn’t Poetry aware about the cp311 wheel there?

Should both wheels have requires-python >=3.10,< 3.12?
 But then if in the future there will wheel for 3.12 I end in the same situation.

**Reply by radoering** [reply]
> Should both wheels have requires-python >=3.10,< 3.12?

Yes.

> But then if in the future there will wheel for 3.12 I end in the same situation.

I'd propose to do a new release if you add Python 3.12 support in the future.

**Reply by abn** [reply]
This is why post releases are useful. You can do a rebuild with just a post release version bump.
